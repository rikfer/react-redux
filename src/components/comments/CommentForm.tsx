import React, { useState } from "react";
import TextInput from "../common/TextInput";
import { Comment } from "../../models/comment/interface";

interface CommentFormProps {
    saving: boolean;
    onSave: (name: string, email: string, body: string) => void;
}

function CommentForm(props: CommentFormProps): JSX.Element {
    const { saving = false, onSave } = props;
    const [name, setName] = useState("");
    const [email, setEmail] = useState("");
    const [body, setBody] = useState("");
    const [errors, setErrors] = useState<any>({});

    const formIsValid = () => {
        const errors: any = {};

        if (!name) errors.name = "Name is required.";
        if (!email) errors.email = "Email is required";
        if (!body) errors.body = "Body is required";

        setErrors(errors);
        // Form is valid if the errors object still has no properties
        return Object.keys(errors).length === 0;
    };

    const cleanForm = () => {
        setName("");
        setEmail("");
        setBody("");
    };

    const saveComment = (event: any) => {
        event.preventDefault();
        console.log(name);
        if (formIsValid()) {
            onSave(name, email, body);
            cleanForm();
        }
    };

    return (
        <form onSubmit={saveComment}>
            <h2>Add a comment</h2>
            {errors.onSave && (
                <div className="alert alert-danger" role="alert">
                    {errors.onSave}
                </div>
            )}
            <TextInput
                name="name"
                label="Name"
                value={name}
                onChange={(e) => setName(e.target.value)}
                error={errors.name}
            />

            <TextInput
                name="email"
                label="Email"
                value={email}
                onChange={(e) => setEmail(e.target.value)}
                error={errors.email}
            />

            <TextInput
                name="body"
                label="Body"
                value={body}
                onChange={(e) => setBody(e.target.value)}
                error={errors.body}
            />

            <button type="submit" disabled={saving} className="btn btn-primary">
                {saving ? "Saving..." : "Save"}
            </button>
        </form>
    );
}

export default CommentForm;
