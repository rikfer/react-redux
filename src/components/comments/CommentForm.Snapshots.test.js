/* eslint-disable no-undef */
import React from "react";
import CommentForm from "./CommentForm";
import renderer from "react-test-renderer";

it("sets submit button label 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <CommentForm
      errors={{}}
      onSave={jest.fn()}
      saving
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label 'Save' when saving is false", () => {
  const tree = renderer.create(
    <CommentForm
      errors={{}}
      onSave={jest.fn()}
      saving={false}
    />
  );

  expect(tree).toMatchSnapshot();
});
