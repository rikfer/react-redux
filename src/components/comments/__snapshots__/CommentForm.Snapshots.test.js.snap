// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Add a comment
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="body"
    >
      Body
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="body"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Add a comment
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="body"
    >
      Body
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="body"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
